r/tests/testthat/test-dplyr.R

```diff
@@ -800,6 +808,21 @@ test_that("type checks with is_*()", {
   )
 })
 
+test_that("type checks on expressions", {
+  expect_dplyr_equal(
+    input %>%
+      transmute(
+        a = is.character(as.character(int)),
```

Comment:
I'm guessing this is the use case that's motivating all of these Expression/Schema issues? I wonder if there's another way to solve it in a more localized way, inside of the is* nse_funcs.

Reply:
Yes—this is a simplistic and somewhat contrived example of the motivating use case. For a more compelling version of it, see ARROW-12055: To make our translation of `is.na()` work consistently with base R, we will need to run different code depending on whether its input has a real number type, and we can't do that if we don't know the type.

Reply:
I created ARROW-13186 for future consideration of ways to implement type determination more cleanly